File: engine/python/framepilot_engine/safety.py
"""Path-sandbox primitives.

WHY: FramePilot is local-first and the AI/agent layer can request file
operations. Per **PRD §18.1 (Local file safety)** and **§18.2 (Agent safety)**,
all file access MUST be sandboxed to the project directory and MUST prevent path
traversal. This module is a real security primitive (not a stub): every file
operation in the engine should resolve user/agent-supplied paths through
:func:`resolve_within` before touching disk.
"""

from __future__ import annotations

from pathlib import Path


class PathTraversalError(Exception):
    """Raised when a candidate path escapes its sandbox base directory.

    This includes ``..`` traversal, absolute paths pointing outside the base,
    and symlink targets that resolve outside the base.
    """


def resolve_within(base: Path, candidate: str) -> Path:
    """Resolve ``candidate`` and guarantee it stays inside ``base``.

    The returned path is fully resolved (symlinks and ``..`` collapsed). If the
    resolved path is not ``base`` itself or a descendant of it, this raises
    :class:`PathTraversalError`.

    Examples::

        >>> base = Path("/projects/demo").resolve()
        >>> resolve_within(base, "assets/clip.mp4")  # doctest: +SKIP
        PosixPath('/projects/demo/assets/clip.mp4')

        >>> resolve_within(base, "../secret.key")  # doctest: +SKIP
        Traceback (most recent call last):
            ...
        framepilot_engine.safety.PathTraversalError: ...

    :param base: The sandbox root directory. Resolved before comparison.
    :param candidate: A relative or absolute path supplied by a caller, the
        user, or the agent. Untrusted input.
    :returns: The resolved, sandbox-checked absolute path.
    :raises PathTraversalError: If the candidate escapes ``base``.
    """
    resolved_base = base.resolve()
    # ``candidate`` may be absolute; Path joining honours that, which is exactly
    # why we must re-check containment after resolving.
    resolved_candidate = (resolved_base / candidate).resolve()

    if resolved_candidate != resolved_base and resolved_base not in resolved_candidate.parents:
        raise PathTraversalError(
            f"Path escapes sandbox. base={resolved_base} candidate={candidate!r} "
            f"resolved={resolved_candidate}"
        )

    return resolved_candidate

File: packages/shared-types/src/safety.ts
/**
 * @framepilot/shared-types/safety — the canonical TS path sandbox.
 *
 * WHY this is shared (node-only) and not per-package: every TS surface that turns
 * a user/agent-supplied path into filesystem IO MUST confine it to a sandbox root
 * and reject traversal (`..`, absolute escapes, and symlinks whose target leaves
 * the sandbox) — PRD §18.1, AGENTS.md §6. This used to live only in the MCP server
 * while the Electron IPC handlers had NO check (Phase 8 security audit finding 1.1).
 * Hoisting it here gives the MCP host and the desktop main process one
 * implementation, and it is a faithful mirror of the engine's
 * `framepilot_engine.safety.resolve_within`, so all three runtimes enforce the same
 * containment rule. See ADR 0025.
 *
 * Imported via the `@framepilot/shared-types/safety` subpath so browser bundles
 * (the renderer) never pull in `node:fs`/`node:path`.
 */
import { realpathSync } from 'node:fs';
import { dirname, resolve, sep } from 'node:path';

/** Raised when a candidate path escapes its sandbox base directory. */
export class PathTraversalError extends Error {
  public constructor(message: string) {
    super(message);
    this.name = 'PathTraversalError';
  }
}

/**
 * Realpath the longest existing ancestor of `target`. The candidate file itself
 * may not exist yet (e.g. a not-yet-saved project), but we still resolve symlinks
 * in its existing prefix — and in the final component when it exists — so a
 * symlinked directory OR a symlinked file cannot smuggle the path out of the
 * sandbox. The non-existing tail was already `..`-collapsed by `resolve()`.
 */
function realExistingPrefix(target: string): string {
  let current = target;
  for (;;) {
    try {
      return realpathSync(current);
    } catch {
      const parent = dirname(current);
      /* v8 ignore next -- base always exists, so we resolve before reaching root. */
      if (parent === current) return current;
      current = parent;
    }
  }
}

/**
 * Resolve `candidate` and guarantee it stays inside `base`.
 *
 * Because `realExistingPrefix` realpaths the deepest existing portion of the
 * resolved target (including the final component when it is an existing symlink),
 * a symlink whose target leaves `base` is rejected even when its parent directory
 * is inside the sandbox.
 *
 * @param base - The sandbox root (e.g. `FRAMEPILOT_PROJECTS_ROOT`). Must exist.
 * @param candidate - An untrusted relative or absolute path from the agent/user.
 * @returns The resolved absolute path (symlinks in its existing prefix collapsed).
 * @throws {PathTraversalError} If the resolved path escapes `base`.
 */
export function resolveWithin(base: string, candidate: string): string {
  const resolvedBase = realpathSync(base);
  // `candidate` may be absolute; resolve() honours that, which is exactly why we
  // must re-check containment after resolving (PRD §18.1).
  const resolved = resolve(resolvedBase, candidate);
  const realPrefix = realExistingPrefix(resolved);
  const contained = realPrefix === resolvedBase || realPrefix.startsWith(resolvedBase + sep);
  if (!contained) {
    throw new PathTraversalError(
      `Path escapes sandbox. base=${resolvedBase} candidate=${JSON.stringify(candidate)} resolved=${resolved}`,
    );
  }
  return resolved;
}
